File: scripts/worker.ts
import "./load-env"
import {
  syncMembers,
  syncTrades,
  backfillTrades,
  syncRankings,
  syncStockPerformance,
  syncFec,
} from "../lib/sync"

const MINUTE = 60_000
const HOUR = 60 * MINUTE

type Job = {
  name: string
  intervalMs: number
  run: () => Promise<unknown>
}

const jobs: Job[] = [
  { name: "members", intervalMs: 24 * HOUR, run: syncMembers },
  { name: "trades", intervalMs: 5 * MINUTE, run: syncTrades },
  { name: "trades-backfill", intervalMs: 24 * HOUR, run: backfillTrades },
  { name: "rankings", intervalMs: 60 * MINUTE, run: syncRankings },
  { name: "stock-performance", intervalMs: 24 * HOUR, run: syncStockPerformance },
  { name: "fec", intervalMs: 24 * HOUR, run: syncFec },
]

const inProgress = new Set<string>()

async function runJob(job: Job): Promise<void> {
  if (inProgress.has(job.name)) {
    console.log(`[worker] ${job.name}: still running, skipping this tick`)
    return
  }
  inProgress.add(job.name)
  const startedAt = Date.now()
  try {
    const result = await job.run()
    const secs = ((Date.now() - startedAt) / 1000).toFixed(1)
    console.log(`[worker] ${job.name}: ok (${secs}s)`, result)
  } catch (error) {
    console.error(
      `[worker] ${job.name}: failed:`,
      error instanceof Error ? error.message : error
    )
  } finally {
    inProgress.delete(job.name)
  }
}

async function main(): Promise<void> {
  console.log("[worker] starting — seeding once, then polling on interval")

  for (const job of jobs) {
    await runJob(job)
  }

  for (const job of jobs) {
    setInterval(() => void runJob(job), job.intervalMs)
  }

  console.log(
    "[worker] scheduled:",
    jobs.map((j) => `${j.name} every ${Math.round(j.intervalMs / MINUTE)}min`).join(", ")
  )
}

void main()

File: backend/worker.py
"""Standalone background worker (port of scripts/worker.ts).

Runs every sync job once at startup, then repeats each on its own interval.

    python worker.py
"""

import asyncio
import logging
import time

from app import config  # noqa: F401  (loads .env.local)
from app.services.sync import (
    backfill_trades,
    sync_fec,
    sync_members,
    sync_rankings,
    sync_stock_performance,
    sync_trades,
)

logging.basicConfig(
    level=logging.INFO, format="%(asctime)s [%(name)s] %(levelname)s: %(message)s"
)
logger = logging.getLogger("worker")

MINUTE = 60
HOUR = 60 * MINUTE

JOBS = [
    {"name": "members", "interval": 24 * HOUR, "run": sync_members},
    {"name": "trades", "interval": 5 * MINUTE, "run": sync_trades},
    {"name": "trades-backfill", "interval": 24 * HOUR, "run": backfill_trades},
    {"name": "rankings", "interval": 60 * MINUTE, "run": sync_rankings},
    {"name": "stock-performance", "interval": 24 * HOUR, "run": sync_stock_performance},
    {"name": "fec", "interval": 24 * HOUR, "run": sync_fec},
]

_in_progress: set[str] = set()


async def run_job(job: dict) -> None:
    if job["name"] in _in_progress:
        logger.info("%s: still running, skipping this tick", job["name"])
        return
    _in_progress.add(job["name"])
    started_at = time.monotonic()
    try:
        result = await job["run"]()
        logger.info(
            "%s: ok (%.1fs) %s", job["name"], time.monotonic() - started_at, result
        )
    except Exception as error:
        logger.error("%s: failed: %s", job["name"], error)
    finally:
        _in_progress.discard(job["name"])


async def schedule(job: dict) -> None:
    while True:
        await asyncio.sleep(job["interval"])
        await run_job(job)


async def main() -> None:
    logger.info("starting — seeding once, then polling on interval")

    for job in JOBS:
        await run_job(job)

    logger.info(
        "scheduled: %s",
        ", ".join(f"{job['name']} every {job['interval'] // MINUTE}min" for job in JOBS),
    )
    await asyncio.gather(*(schedule(job) for job in JOBS))


if __name__ == "__main__":
    asyncio.run(main())
